packages/react-aria-components/src/GridList.tsx

```diff
@@ -97,38 +104,57 @@ interface GridListInnerProps<T extends object> {
 function GridListInner<T extends object>({props, collection, gridListRef: ref}: GridListInnerProps<T>) {
   let {dragAndDropHooks, keyboardNavigationBehavior = 'arrow', layout = 'stack'} = props;
   let {CollectionRoot, isVirtualized, layoutDelegate, dropTargetDelegate: ctxDropTargetDelegate} = useContext(CollectionRendererContext);
+
+  // Unscope the node keys for convenient access.
+  let disabledKeys = useMemo(() => !props.disabledKeys ? props.disabledKeys : [...props.disabledKeys].map(key => `${props.id}:${key}`), [props.disabledKeys, props.id]);
+  let selectedKeys = useMemo(() => !props.selectedKeys || props.selectedKeys === 'all' ? props.selectedKeys : [...props.selectedKeys].map(key => `${props.id}:${key}`), [props.selectedKeys, props.id]);
+  let defaultSelectedKeys = useMemo(() => !props.defaultSelectedKeys || props.defaultSelectedKeys === 'all' ? props.defaultSelectedKeys : [...props.defaultSelectedKeys].map(key => `${props.id}:${key}`), [props.defaultSelectedKeys, props.id]);
+  let onSelectionChange = useMemo(() => !props.onSelectionChange ? props.onSelectionChange : (keys: Selection) => {
+    props.onSelectionChange?.(new Set([...keys].map(key => typeof key === 'string' ? key.replace(`${props.id}:`, '') : key)));
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.onSelectionChange, props.id]);
+  let onAction = useMemo(() => !props.onAction ? props.onAction : (key: Key) => {
+    props.onAction?.(typeof key === 'string' ? key.replace(`${props.id}:`, '') : key);
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.onAction, props.id]);
```

Comment:
One thing I just noticed is that this unscoping doesn't work with the Virtualizer wrapped GridList. It seems like the GridList's collection gets built with one instance of the autogenerated id, but then another render happens where the GridList generates another scope id but the collection itself doesn't get regenerated. This results in the above `props.id` mismatching with the collection's node key prefix. Poking around to see if I can figure out why that is

Reply:
Sounds like strict mode maybe? Will habe to check in depth tmrw.

Reply:
Good catch, just got around to take a proper look. I pushed a fix to remote 👍 

The issue wasn't with `useId` but rather with the drag and drop hooks generating an id without taking the `defaultId` into account. That also explains why you only observed it with `GridList` as all other virtualized examples simple didn't have drag and drop implemented 😄 

I opted to override the id in `mergeProps` of the component, but longterm might be wise to think about forwarding the `default` to the drag and drop hooks?

Reply:
Ah, nice catch! I completely forgot that the drop hooks generate an `id` of their own, was so confused how some autogenerated id was making it though. Unfortunately this means I've found even more problematic areas in the code 😅.

1. I think it is a good idea to forward the GridList's `id` to GridList's call o `useDroppableCollection`  so the weak map id [here](https://github.com/adobe/react-spectrum/blob/4b971490601c43f6fffb663cfdf7081ddc77c04d/packages/%40react-aria/dnd/src/useDroppableCollection.ts#L748-L749) is consistent. It actually already works at the moment due to `mergeProps` deduping the ids and causing all other instances of `useId` to use the last one provided ([aka via this call of mergeProps)](https://github.com/adobe/react-spectrum/blob/8e9dfb81d330aab2ce1e322443e3b4013c149003/packages/react-aria-components/src/GridList.tsx#L383), but it would be best not to rely on this IMO.

2. I noticed that the reordering in the Virtualized Gridlist story is broken because the selectionManager/collection has the prefixed `id` for the items (which is expected) and will use those prefixed `id` as [the dragged keys when drag starts](https://github.com/adobe/react-spectrum/blob/4b971490601c43f6fffb663cfdf7081ddc77c04d/packages/%40react-aria/dnd/src/useDraggableItem.ts#L77-L79) (see [here](https://github.com/adobe/react-spectrum/blob/4b971490601c43f6fffb663cfdf7081ddc77c04d/packages/%40react-stately/dnd/src/useDraggableCollectionState.ts#L73-L85) for how it is pulling them out of the `keyMap`). These are stored globally and used in [useDroppableCollection](https://github.com/adobe/react-spectrum/blob/4b971490601c43f6fffb663cfdf7081ddc77c04d/packages/%40react-aria/dnd/src/useDroppableCollection.ts#L99) in things like `onReorder`/etc, resulting in logic like https://github.com/adobe/react-spectrum/blob/4b971490601c43f6fffb663cfdf7081ddc77c04d/packages/react-aria-components/stories/GridList.stories.tsx#L130-L136 no longer working because the `list` has the user's original non prefixed `id` but `e.keys` has the prefixed `ids` and thus we have a mismatch. If we store the `prefix` and as part of the `collection`/`SelectionManager` via `BaseCollection` like you mentioned before, we could perhaps strip off the prefix and hopefully fix this

However, I'm now a bit more worried since this class of issue is due to our original implemented assumption that the `ids` stored in the `collection`/`SelectionManager` matches the original `ids` format provided by the user. This assumption is likely to exist in many other hooks, and I'm quite troubled by the fact that the tests ran clean...

Reply:
Perhaps we should step back and instead have the collection id scoping only applied to the generated DOM data-key rather than affecting the keys at collection/SelectionManger level.

Reply:
haha, that's what i just had in mind aswell. pushed it to remote 👍

Reply:
Btw, funnily enough your call about `useId` caused me to unearth a tricky little bug in it. 

When a component has 2 ids for some reason (e.g. because of drag & drop hooks), merging these ids in the order `id1 -> id2 -> id1` causes the component to infinitely re-render when in strict mode 😄. 

Not sure whether it's worth to fix that, but could be untangled by `mergeProps` perhaps.